### ISSUE TITLE
Disambiguate server in OpenAPI specs with multiple server objects

### ISSUE THREAD

**Issue by kevinteg** [serverlessworkflow/specification#367]
**What would you like to be added**:

It is possible to express multiple server objects in Open API spec:
https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#oasServers

It is also possible to express the serverUrl as a relative path (to the spec location):
https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#serverUrl

I would like the ability to specify a hostname in the rest function definition in cases where the server url is ambiguous (or incorrect).  I know it is common to have a single spec for many endpoints and the spec doesn't have the list of all of them.

**Why is this needed**:
Runtimes will need additional information to determine the rest endpoint when the Open API spec does not specify a single absolute URL.

**Comment by tsurdilo** [comment]
@Neuroglia @mswiderski @manuelstein @jorgenj @ricardozanini @antmendoza ^^ 

Currently we enforce operationId to be specified on the path even tho its an optional parameter. 

To me it seems that "servers" is optional as well. The one thing I am trying to understand is:
it looks as the servers prop does not have a unique identifier it has url and description only. It does look as you can
set uniqueness via "variables" but I am not 100% sure if thats really used for that or something else?

So question is can we enforce that openapi def has the servers prop? How should be deal with identifying a server def 
by some unique value rather than having to hard-code url end-points inside workflow definition?

If we don't enforce, what should we do about use case that @kevinteg describes where openapi def does not provide servers info?

**Comment by mswiderski** [comment]
in my opinion spec should be agnostic of the actual location of the service that is being invoked. It is more of a runtime to be able to connect these together. In case where the servers are given in open api definition then it should be considered as default service url. Regardless if it is defined in open api or not it should always be easily to override it e.g. environment variable as that will give you simple way to test your workflow on various environments where the service you're invoking is in different servers or environments (dev, test, qa, prod etc). If you are interested in example of that I can reference one in the runtime implementation I work with.

**Comment by Neuroglia** [comment]
@tsurdilo  I totally agree with @mswiderski: the spec should not bother with that, it should (keep) be(ing) agnostic.
In Synapse runtime, I actually try to run all defined servers one after the other, in case I received a 503.
If servers is not specified (which IMHO should be enforced by OpenApi), I implicitly consider the host of the document to be the default server.
Other cases can be easily covered by runtimes with, as @mswiderski said, Environment Variables, config maps, secrets, ...

**Comment by jorgenj** [comment]
> in my opinion spec should be agnostic of the actual location of the service that is being invoked

100% agree, there are use-cases where the spec will be served from a completely different location than the service.  The openapi spec says that when no serverObject is defined, that the root of the url for the spec is what should be used as the server.

> In Synapse runtime, I actually try to run all defined servers one after the other, in case one fails with a 503. In other words, I consider additional servers as backups.

This may not be a good idea, given that the openapi specification gives the following example (so clearly it appears to be a use-case they expect to be used):

https://swagger.io/specification/#server-object

![image](https://user-images.githubusercontent.com/12246/117182087-c65ce180-ad8a-11eb-9908-fbc6a91dec5e.png)

> Other cases can be easily covered by runtimes with, as @mswiderski said, Environment Variables, config maps, secrets, ...

There are cases where it's critical to the correctness of a workflow that the right server be targeted, externalizing this selection logic outside of the workflow seems counter-productive to the portability goals of this serverless workflow spec.   The server selection (in the case where there are many servers defined) can potentially have dangerous consequences if done incorrectly, I agree that some context needs to be provided to aid that selection but I don't see how a workflow can be considered portable if that selection criteria isn't included in the workflow definition.

**Comment by tsurdilo** [comment]
@jorgenj my thinking is that hard-coding urls in main DSL props is much less portable than:

1) using environmental vars/whatever
2) passing in server id/url as workflow data input
3) using workflow annotations property to map on runtime level

I think you can always add hard-coded urls inside the function definition metadata if nothing else works for your runtime

**Comment by jorgenj** [comment]
> @jorgenj my thinking is that hard-coding urls in main DSL props is much less portable than:

This definition of "portable" is....difficult to comprehend...

> 
> 1. using environmental vars/whatever

How is the workflow definition supposed to indicate which environment variables to use as the server for each openapi function?   The runtimes can just each make up some convention they follow, but this hardly sounds more "portable".

> 2. passing in server id/url as workflow data input
> 3. using workflow annotations property to map on runtime level

As a user, in my workflow definition how do I tell the runtime that for this function invocation to pull the server from the workflow input data or annotations?

> 
> I think you can always add hard-coded urls inside the function definition metadata if nothing else works for your runtime (which in itself is more portable as other runtimes dont have to use it)

Isn't the server a fundamental piece of information needed to make an API call?  I thought that .metadata was not allowed to contain anything that affects the execution of the workflow, surely if we don't know what server to call then the workflow can't successfully execute, ie; the server clearly affects the 'runnability' of a workflow.

**Comment by tsurdilo** [comment]
@jorgenj 
when you write your Java code, do you consider hard-coding String values inside your mission-critical code a good practice? Thats kinda what you are asking us to do here I think :)

**Comment by tsurdilo** [comment]
>Isn't the server a fundamental piece of information needed to make an API call? I thought that .metadata was not allowed to contain anything that affects the execution of the workflow, surely if we don't know what server to call then the workflow can't successfully execute, ie; the server clearly affects the 'runnability' of a workflow.

When you write workflows you need to consider failure and handle it in your workflow definition. 

When we say for metadata that it should not include things that  affects control flow logic,
we mean that those values should not affect the determinism of the workflow definition.

For example, if workflow metadata overwrites workflow ids, versions, state names, transitions, things that cause the defined execution paths (including paths that handle errors during execution) to be different than whats defined in the workflow definition then they are not allowed. Hope this helps, we should add some more info to the spec document about this possibly.

**Comment by jorgenj** [comment]
> @jorgenj
> when you coding, do you consider hard-coding an URL string value in your mission-critical code a good practice? Thats kinda what you are arguing to do here I think :)

This is a gross misrepresentation of what I'm asking.

**Comment by kevinteg** [comment]
My take on this is that if the CNCF workflow spec wants to use OpenAPI for function references, it should consider making it easier for users who may not have control over the OpenAPI spec.  In a lot of situations, users will be making API calls to external services or to globally diverse services where there may be hundreds of endpoints for the same service where maintaining one spec per endpoint is unfeasible.  Providing a mechanism to influence this will likely save users from a lot of grief (either influencing external services to modify their specs or host one spec per endpoint, or modifying and hosting a copy of the CNCF compatible Open API spec on their own).

**Comment by tsurdilo** [comment]
@kevinteg 👍 thanks for the inputs!

To me this seems as a limitation of OpenApi server definition that we need to possibly address and cover for?
We should be able to completely rely on OpenApi for invocation details, including servers. 
By us defining a hostname parameter does that mean our DSL overwrites the instructions defined by the people who write their APIs? Also note that our function definitions are reusable, meaning that you can define them in a separate json/yaml file and use them in multiple workflow definitions, where different ones might need a different server url to "successfully execute".

So to summarize, we have a couple of runtimes which are able to overcome this possible limitation with openapi server definitions in different ways, 
we also talked about what can be done currently with our dsl , such as use of function definition metadata, annotation props as well as combination of workflow data input + metadata that might be feasible solutions to define a server url.

It was suggested in this Issue to add a new "hostname" parameter containing a specific URL of the server which has to be invoked for that particular function definition. 

We will definitely add this as a discussion item in one of the upcoming team meetings to get more opinions from the team.

I would like to move this discussion to team meetings please. Alternatively feel free to continue it on our discussion board: https://github.com/serverlessworkflow/specification/discussions

**Comment by mswiderski** [comment]
last final comment from my end - hardcoding in the workflow definition any information that are environment specific has always been considered as bad practice. That be urls, usernames, passwords and so on. This makes the workflow definition (regardless of the dsl it's expressed in) not portable between environments and fragile when it comes to potential server/users changes.

In addition, this does not refer to OpenAPI only as the same is for events. Would you like to also include in the workflow where events coming from or are sent to? Like Kafka server, AMQP etc. It has way more than just the location of the server or broker that the runtime will have to deal with and like it or not it will not be portable as different technologies and programming languages deal with that differently. So trying to enforce this on the workflow definition level is not going to make it better anyway. OpenAPI has it rather nicely covered but yeah some of these are optional but in many cases they are available and if not they can be provided as overrides.
When it comes to multiple servers defined in the spec, it is still up to the runtime and user to select properly which server url is going to be used as sometimes it is completely fine to use round robin calls to all server lists (assuming they are there for load balancing) sometimes they are not as they will refer to different environments of the same service. So there is no way this could be handled on the spec level as you would end up with exact same thing - defining multiple urls in workflow and no way to know which one to use.

**Comment by Neuroglia** [comment]
> This may not be a good idea, given that the openapi specification gives the following example (so clearly it appears to be a use-case they expect to be used):

@jorgenj I know about OpenApi spec defining the possible use of multiple servers, I'm fine with it. I just fail to understand why implicitly considering additional servers as 'back-ups' is a bad idea according to you. 

If you mean that using a round-robin approach to servers that might not have been declared for this purpose (but rather to centralize the delcaration different independant servers), I feel you. That's why the round-robin approach in case of 503 is a default, optional behavior. You could definilty change that to something like: AtLeastOne, AtMostN, All, RoundRobin, Auto
Fallback

**Comment by Neuroglia** [comment]
Again, I totally agree with @mswiderski last comment.

Adding support for this feature (which IMHO is a hornet nest) will just complexify the spec and step over responsabilities of third party specs.

> Would you like to also include in the workflow where events coming from or are sent to? 

I really hope we never do such a thing. At best, we should use the AsyncApi spec to do that. It's again not the responsibility of SW to define this and possibly conflict with others.

In addition, as you said, there are so many use-cases for using multiple servers in OpenApi that trying to address the said issue will eventually just spawn new ones.

**Comment by kevinteg** [comment]
Thanks for the replies everyone!  I'll add my final reply too (it sounds like this proposal is rejected so no point carrying further in a separate discussion).

I get the impression that for a lot of people, the runtime owners are in the same group/company as the users and that they may be in a position to support specific user cases by configuring the hosting infrastructure in various ways (config maps, environment variables).  I know there are a lot of options for handling this at the runtime if we control the endpoints but I'm specifically thinking of use cases where the user/customer wants to call an external service when they don't own the Open API spec.

I suggested this features as a way of making it crystal clear what the endpoint is if the OpenAPI spec the user provides doesn't specify.  In theory I completely agree that it would be nice to change the OpenAPI specs to provide a single endpoint, but in practice I don't see that happening.

To the question of whether I think this should be in other fields: probably if the spec we use doesn't provide enough information for the runtime to perform the operation.  I didn't see the harm in specifically supplementing the OpenAPI spec in functions of "rest" type.  For events, probably not - I think that would be defined by the runtime (but who knows- maybe it is desirable for the runtime to subscribe to customer defined event sources).

It sounds like the answer is to use metadata but I got the impression it shouldn't affect workflow execution.
